core/tests/interfaces.rs

```diff
@@ -0,0 +1,279 @@
+        "Furniture",
+    );
+
+    let query = "query { leggeds { legs } }";
+
+    let res = insert_and_query(subgraph_id, schema, vec![animal, furniture], query).unwrap();
+    assert!(res.errors.is_none());
+    assert_eq!(
+        format!("{:?}", res.data.unwrap()),
+        "Object({\"leggeds\": List([Object({\"legs\": Int(Number(3))}), Object({\"legs\": Int(Number(3))})])})"
+    )
+}
+
+#[test]
+fn reference_interface() {
+    let subgraph_id = "ReferenceInterface";
+    let schema = "type Leg @entity { id: ID! }
+                  interface Legged { leg: Leg }
+                  type Animal implements Legged @entity { id: ID!, leg: Leg }";
+
+    let query = "query { leggeds { leg { id } } }";
+
+    let leg = (Entity::from(vec![("id", Value::from("1"))]), "Leg");
+    let animal = (
+        Entity::from(vec![("id", Value::from("1")), ("leg", Value::from("1"))]),
+        "Animal",
+    );
+
+    let res = insert_and_query(subgraph_id, schema, vec![leg, animal], query).unwrap();
+
+    assert!(res.errors.is_none());
+    assert_eq!(
+        format!("{:?}", res.data.unwrap()),
+        "Object({\"leggeds\": List([Object({\"leg\": Object({\"id\": String(\"1\")})})])})"
+    )
+}
+
+#[test]
+fn conflicting_implementors_id() {
+    let subgraph_id = "ConflictingImplementorsId";
+    let schema = "interface Legged { legs: Int }
+                  type Animal implements Legged @entity { id: ID!, legs: Int }
+                  type Furniture implements Legged @entity { id: ID!, legs: Int }
+                  ";
+
+    let animal = (
+        Entity::from(vec![("id", Value::from("1")), ("legs", Value::from(3))]),
+        "Animal",
+    );
+    let furniture = (
+        Entity::from(vec![("id", Value::from("1")), ("legs", Value::from(3))]),
+        "Furniture",
+    );
+
+    let query = "query { leggeds { legs } }";
+
+    let res = insert_and_query(subgraph_id, schema, vec![animal, furniture], query);
+    assert_eq!(
+        res.unwrap_err().to_string(),
+        "tried to set entity of type `Furniture` with ID `1` but an entity of type `Animal`, \
+         which has an interface in common with `Furniture`, exists with the same ID"
+    );
+}
+
+#[test]
+fn derived_interface_relationship() {
+    let subgraph_id = "DerivedInterfaceRelationship";
+    let schema = "interface ForestDweller { id: ID!, forest: Forest }
+                  type Animal implements ForestDweller @entity { id: ID!, forest: Forest }
+                  type Forest @entity { id: ID!, dwellers: [ForestDweller]! @derivedFrom(field: \"forest\") }
+                  ";
+
+    let forest = (Entity::from(vec![("id", Value::from("1"))]), "Forest");
+    let animal = (
+        Entity::from(vec![("id", Value::from("1")), ("forest", Value::from("1"))]),
+        "Animal",
+    );
+
+    let query = "query { forests { dwellers { id } } }";
+
+    let res = insert_and_query(subgraph_id, schema, vec![forest, animal], query);
```

Comment:
Having other things in the store for the tests is a good idea.

I think `derivedFrom` should be forbidden in interfaces, the interface doesn't care if a field is derived or not. In fact I hacked a bit on `derivedFrom` validation, I have a follow up PR for that.

Reply:
While it would be nice to have unrelated things in the store for tests, kind of like a fuzzing, it's hard to do that in a general way and I didin't want to clutter all tests with that. So I just added a new `two_interfaces` test for a case like this.

Reply:
Cool, that's something 😄

Reply:
True, whether a relationship is `@derivedFrom` is more of an implementation detail in the interface-implementing types. Mu only thought here is that if these types implement a field different, then resolving their fields would have to take that into account.